File: webpreview.py
import sys
import cairo, gi

DEFAULT_WIDTH = 800
DEFAULT_HEIGHT = 1024
MINIMUM_HEIGHT = 50
TIMEOUT_DELAY = 10 * 1000

USE_GTK2 = True

if USE_GTK2:
	gi.require_version('Gtk', '2.0')
	gi.require_version('Gdk', '2.0')
	gi.require_version('WebKit', '1.0')
from gi.repository import GLib, Gtk, Gdk, WebKit

if len(sys.argv) != 3:
	print 'usage: %s url filename'%sys.argv[0]
	sys.exit(-1)

uri, filename = sys.argv[1:]

for scheme in ('http://', 'https://', 'ftp://', 'file://'):
	if uri.startswith(scheme):
		break
else:
	uri = 'http://' + uri

window = Gtk.OffscreenWindow()
window.props.default_width = DEFAULT_WIDTH
view = WebKit.WebView()
settings = view.get_settings()
settings.props.auto_resize_window = True
settings.props.enable_plugins = False
view.set_settings(settings)
window.add(view)

def capture_window(window):
	w, h = window.get_size()
	if h < MINIMUM_HEIGHT:
		window.set_default_size(w, DEFAULT_HEIGHT)
		view.reload()
		return
	src = Gdk.cairo_create(window.get_window()).get_target()
	src.write_to_png(filename)
	Gtk.main_quit()

def load_cb(webview, pspec, window):
	if webview.get_load_status() == WebKit.LoadStatus.FINISHED:
		capture_window(window)

view.connect('notify::load-status', load_cb, window)
view.load_uri(uri)
window.show_all()

GLib.timeout_add(TIMEOUT_DELAY, Gtk.main_quit)

Gtk.main()

File: webpreview.js
const Mainloop = imports.mainloop;
const Lang = imports.lang;

DEFAULT_WIDTH = 800
DEFAULT_HEIGHT = 1024
MINIMUM_HEIGHT = 50
TIMEOUT_DELAY = 10 * 1000

USE_GTK2 = true

if (USE_GTK2) {
	imports.gi.versions.Gtk = '2.0'
	imports.gi.versions.Gdk = '2.0'
	imports.gi.versions.WebKit = '1.0'
}

const Gdk = imports.gi.Gdk
const Gtk = imports.gi.Gtk
const WebKit = imports.gi.WebKit

function WebPreview(url, filename) {
    this._init(url, filename);
}

WebPreview.prototype = {
	_init : function(url, filename) {
		this.win = new Gtk.OffscreenWindow()
		this.win.default_width = DEFAULT_WIDTH
		this.view = new WebKit.WebView()
		settings = this.view.get_settings()
		settings.auto_resize_window = true
		settings.enable_plugins = false
		this.view.set_settings(settings)
		this.win.add(this.view)

		this.view.connect('notify::load-status', Lang.bind(this, this.load_cb))

		this.view.load_uri(uri)
		this.win.show_all()
	},

	capture_window : function() {
		[w, h] = this.win.get_size()
		if (h < MINIMUM_HEIGHT) {
			this.win.set_default_size(w, DEFAULT_HEIGHT)
			this.view.reload()
			return
		}
		src = Gdk.cairo_create(this.win.get_window()).getTarget()
		src.writeToPNG(filename)
		Gtk.main_quit()
	},

	load_cb : function(o, pspec) {
		if (this.view.get_load_status() == WebKit.LoadStatus.FINISHED) {
			this.capture_window(this.win)
		}
	}
}

function main(args) {

	[uri, filename] = args

	schemes = ['http://', 'https://', 'ftp://', 'file://']
	has_invalid_scheme = true
	for (let i in schemes) {
		if (uri.search(schemes[i]) == 0) {
			has_invalid_scheme = false
			break
		}
	}

	if (has_invalid_scheme) {
		uri = 'http://' + uri
	}

	Gtk.init(null, 0);

	preview = new WebPreview(uri, filename)
	Mainloop.timeout_add(TIMEOUT_DELAY, Gtk.main_quit)

	Gtk.main()
}

if (ARGV.length == 2) {
	main(ARGV)
} else {
  print('usage: webpreview url filename')
}


